File: graph_dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

// Weighted Graph ke liye hum pair use karenge: {neighbor_node, weight}
void addEdge(vector<pair<int, int>> adj[], int u, int v, int weight) {
    adj[u].push_back({v, weight});
    adj[v].push_back({u, weight}); // Undirected graph
}

void dijkstra(vector<pair<int, int>> adj[], int V, int src) {
    // Min-Heap priority queue: hamesha minimum distance wali node top par rahegi
    // Format: pair<distance, node>
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    vector<int> dist(V, INT_MAX);

    // Source node setup
    dist[src] = 0;
    pq.push({0, src});

    while (!pq.empty()) {
        int d = pq.top().first;
        int node = pq.top().second;
        pq.pop();

        // Agar hume isse behtar rasta pehle hi mil chuka hai, toh skip karein
        if (d > dist[node]) continue;

        // Padosiyo ko check karein
        for (auto edge : adj[node]) {
            int neighbor = edge.first;
            int weight = edge.second;

            // Relaxation condition
            if (dist[node] + weight < dist[neighbor]) {
                dist[neighbor] = dist[node] + weight;
                pq.push({dist[neighbor], neighbor});
            }
        }
    }

    // Output Print karein
    cout << "Dijkstra Shortest Distance from Source (Node " << src << "):" << endl;
    for (int i = 0; i < V; i++) {
        cout << "To Node " << i << " -> Minimum Cost: " << dist[i] << endl;
    }
}

int main() {
    cout << "--- DSA Masterclass: Dijkstra's Algorithm ---" << endl << endl;

    int V = 4;
    vector<pair<int, int>> adj[V];

    // Connections with weights: addEdge(adj, u, v, weight)
    addEdge(adj, 0, 1, 4); // 0 se 1 jaane ka karcha 4 hai
    addEdge(adj, 0, 2, 1); // 0 se 2 jaane ka karcha 1 hai
    addEdge(adj, 2, 1, 2); // 2 se 1 jaane ka karcha 2 hai
    addEdge(adj, 1, 3, 3); // 1 se 3 jaane ka karcha 3 hai
    addEdge(adj, 2, 3, 8); // 2 se 3 jaane ka karcha 8 hai

    dijkstra(adj, V, 0);

    cout << "---------------------------------------------" << endl;
    return 0;
}

File: graph_dijkstra.py
import heapq

class GraphDijkstra:
    def __init__(self):
        self.graph = {}

    # वेटेड ग्राफ के लिए कनेक्शन जोड़ना (u से v का वजन/दूरी w है)
    def add_edge(self, u, v, weight):
        if u not in self.graph:
            self.graph[u] = []
        if v not in self.graph:
            self.graph[v] = []
            
        self.graph[u].append((v, weight))
        self.graph[v].append((u, weight))  # Undirected Graph

    # डाईकस्ट्रा एल्गोरिदम मुख्य फंक्शन
    def dijkstra(self, start_node):
        # शुरुआत में सभी नोड्स की दूरी Infinity (असीम) मान लेते हैं
        distances = {node: float('inf') for node in self.graph}
        distances[start_node] = 0
        
        # मिनिमम हीप (Priority Queue) सबसे छोटे रास्ते को पहले प्रोसेस करने के लिए
        min_heap = [(0, start_node)]  # (distance, node)

        while min_heap:
            current_distance, current_node = heapq.heappop(min_heap)

            # अगर हमें पहले से ही कोई छोटा रास्ता मिल चुका है, तो आगे बढ़ें
            if current_distance > distances[current_node]:
                continue

            # पड़ोसी नोड्स और उनकी दूरी को चेक करें
            for neighbor, weight in self.graph[current_node]:
                distance = current_distance + weight

                # अगर नया रास्ता पुराना रास्ते से छोटा है, तो अपडेट करें
                if distance < distances[neighbor]:
                    distances[neighbor] = distance
                    heapq.heappush(min_heap, (distance, neighbor))

        # रिजल्ट प्रिंट करें
        print(f"--- Shortest distances from node {start_node} ---")
        for node, dist in distances.items():
            print(f"Distance to node {node} : {dist}")


# --- Main Driver Code ---
if __name__ == "__main__":
    g = GraphDijkstra()
    
    # ग्राफ में नोड्स और उनके बीच की दूरी (Weight) जोड़ रहे हैं
    g.add_edge(0, 1, 4)
    g.add_edge(0, 2, 1)
    g.add_edge(2, 1, 2)
    g.add_edge(1, 3, 5)
    g.add_edge(2, 3, 8)

    # नोड 0 से सभी नोड्स की सबसे छोटी दूरी निकालें
    g.dijkstra(0)